### ISSUE TITLE
a error for dynamic partition pruning?

### ISSUE THREAD

**Issue by liruto** [greenplum-db/gpdb#271]
Due to gp_dynamic_partition_pruning setted on/off, different result sets were gained.
A case as follow:

```
mycluster=# drop table t1,t2;
DROP TABLE
mycluster=# drop sequence cs1;
DROP SEQUENCE
mycluster=# create table t1(tc1 int, tc2 int) distributed by (tc1) partition by range(tc1) (partition p1 end(10), partition p2 end(20));
NOTICE:  CREATE TABLE will create partition "t1_1_prt_p1" for table "t1"
NOTICE:  CREATE TABLE will create partition "t1_1_prt_p2" for table "t1"
CREATE TABLE
mycluster=# create table t2(tc1 int, tc2 int) distributed randomly;
CREATE TABLE
mycluster=# insert into t1 values(2,2);
INSERT 0 1
mycluster=# insert into t2 values(2,2);
INSERT 0 1
mycluster=# create sequence cs1;
CREATE SEQUENCE
mycluster=# select nextval('cs1');
 nextval 
---------
       1
(1 row)

mycluster=# set gp_dynamic_partition_pruning=on;
SET
mycluster=# explain select * from t1, t2 where t1.tc1 = t2.tc1 and t2.tc1 = nextval('cs1');
                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Gather Motion 4:1  (slice3; segments: 4)  (cost=2.08..13.94 rows=10 width=16)
   ->  Hash Join  (cost=2.08..13.94 rows=3 width=16)
         Hash Cond: public.t1.tc1 = public.t2.tc1
         InitPlan  (slice4)
           ->  Aggregate  (cost=1.02..1.03 rows=1 width=32)
                 ->  Gather Motion 4:1  (slice2; segments: 4)  (cost=0.00..1.01 rows=1 width=8)
                       ->  Seq Scan on t2  (cost=0.00..1.01 rows=1 width=8)
                             Filter: tc1 = nextval('cs1'::regclass)
         ->  Append  (cost=0.00..9.61 rows=216 width=8)
               ->  Result  (cost=0.00..9.61 rows=216 width=8)
                     One-Time Filter: 41376::oid = ANY ($0)
                     ->  Seq Scan on t1_1_prt_p1 t1  (cost=0.00..9.61 rows=216 width=8)
               ->  Result  (cost=0.00..0.00 rows=1 width=8)
                     One-Time Filter: 41403::oid = ANY ($0)
                     ->  Seq Scan on t1_1_prt_p2 t1  (cost=0.00..0.00 rows=1 width=8)
         ->  Hash  (cost=1.03..1.03 rows=1 width=8)
               ->  Redistribute Motion 4:4  (slice1; segments: 4)  (cost=0.00..1.03 rows=1 width=8)
                     Hash Key: public.t2.tc1
                     ->  Seq Scan on t2  (cost=0.00..1.01 rows=1 width=8)
                           Filter: tc1 = nextval('cs1'::regclass)
 Settings:  gp_dynamic_partition_pruning=on
(21 rows)

mycluster=# select * from t1, t2 where t1.tc1 = t2.tc1 and t2.tc1 = nextval('cs1');
 tc1 | tc2 | tc1 | tc2 
-----+-----+-----+-----
(0 rows)

mycluster=# drop sequence cs1;
DROP SEQUENCE
mycluster=# create sequence cs1;
CREATE SEQUENCE
mycluster=# select nextval('cs1');
 nextval 
---------
       1
(1 row)

mycluster=# set gp_dynamic_partition_pruning=off;
SET
mycluster=# explain select * from t1, t2 where t1.tc1 = t2.tc1 and t2.tc1 = nextval('cs1');
                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Gather Motion 4:1  (slice2; segments: 4)  (cost=1.05..12.91 rows=10 width=16)
   ->  Hash Join  (cost=1.05..12.91 rows=3 width=16)
         Hash Cond: public.t1.tc1 = t2.tc1
         ->  Append  (cost=0.00..9.61 rows=216 width=8)
               ->  Seq Scan on t1_1_prt_p1 t1  (cost=0.00..9.61 rows=216 width=8)
               ->  Seq Scan on t1_1_prt_p2 t1  (cost=0.00..0.00 rows=1 width=8)
         ->  Hash  (cost=1.03..1.03 rows=1 width=8)
               ->  Redistribute Motion 4:4  (slice1; segments: 4)  (cost=0.00..1.03 rows=1 width=8)
                     Hash Key: t2.tc1
                     ->  Seq Scan on t2  (cost=0.00..1.01 rows=1 width=8)
                           Filter: tc1 = nextval('cs1'::regclass)
 Settings:  gp_dynamic_partition_pruning=off
(12 rows)

mycluster=# select * from t1, t2 where t1.tc1 = t2.tc1 and t2.tc1 = nextval('cs1');
 tc1 | tc2 | tc1 | tc2 
-----+-----+-----+-----
   2 |   2 |   2 |   2
(1 row)
```

In my opinion, parameters for performance optimization couldn't bring impact on the accuracy of results.Maybe, some extra restrictions at 'MatchPartitionJoinPattern' or somewhere else are need. 
Here is that, i think, righttree can't contain volatile function 'nextval'.

**Comment by asimrp** [comment]
@xinzweb, @vraghavan78, can someone please take a look at this?  Two plans for the same query are yielding different results.

**Comment by renkemingithub** [comment]
of course , it is different.   nextval('cs1') is changed. not same value.

**Comment by vraghavan78** [comment]
Thanks @liruto sorry for the delayed response. I have created a story to look into it. Will update the issue once we have done our root cause analysis.

**Comment by hlinnaka** [comment]
This is essentially the same issue that I reported much later in issue #2116. I didn't notice this existing issue at the time. It was fixed by commit e378d84bf3.
